lib/bitcrowd_ecto/changeset.ex

```diff
@@ -112,4 +112,137 @@ defmodule BitcrowdEcto.Changeset do
       add_error(changeset, field, "is not a valid url", validation: :format)
     end
   end
+
+  @doc """
+  Validates a field timestamp to be in the past, if present
+  """
+  @doc since: "0.6.0"
+  @spec validate_past_datetime(Ecto.Changeset.t(), atom, DateTime.t()) :: Ecto.Changeset.t()
+  def validate_past_datetime(changeset, field, now \\ DateTime.utc_now()) do
+    datetime = get_change(changeset, field)
+
+    if datetime && DateTime.compare(now, datetime) == :lt do
+      add_error(changeset, field, "must be in the past", validation: :date_in_past)
+    else
+      changeset
+    end
+  end
+
+  @doc """
+  Validates a field timestamp to be in the future, if present
+  """
+  @doc since: "0.6.0"
+  @spec validate_future_datetime(Ecto.Changeset.t(), atom, DateTime.t()) :: Ecto.Changeset.t()
+  def validate_future_datetime(changeset, field, now \\ DateTime.utc_now()) do
+    datetime = get_change(changeset, field)
+
+    if datetime && DateTime.compare(now, datetime) != :lt do
+      add_error(changeset, field, "must be in the future", validation: :date_in_future)
+    else
+      changeset
+    end
+  end
+
+  @doc """
+  Validates a field timestamp to be after the given one
+  """
+  @doc since: "0.6.0"
+  @spec validate_datetime_after(Ecto.Changeset.t(), atom, DateTime.t(), fun) :: Ecto.Changeset.t()
+  def validate_datetime_after(
+        changeset,
+        field,
+        reference_datetime,
+        formatter \\ &DateTime.to_string/1
+      ) do
+    datetime = get_change(changeset, field)
+
+    if datetime && DateTime.compare(reference_datetime, datetime) != :lt do
+      add_error(
+        changeset,
+        field,
+        "must be after #{formatter.(reference_datetime)}",
+        validation: :datetime_after
+      )
+    else
+      changeset
+    end
+  end
+
+  @doc """
+  Validates two date fields to be a date range, so if both are set the first field has to be
+  before the second field. The error is placed on the later field.
+  """
+  @doc since: "0.6.0"
+  @spec validate_date_order(Ecto.Changeset.t(), atom, atom, list(atom)) :: Ecto.Changeset.t()
+  def validate_date_order(changeset, from_field, until_field, valid_orders \\ [:lt, :eq]) do
+    validate_order(
+      changeset,
+      from_field,
+      until_field,
+      :date_order,
+      &Date.compare/2,
+      valid_orders
+    )
+  end
+
+  @doc """
+  Validates two datetime fields to be a time range, so if both are set the first has to be before
+  the second field. The error is placed on the later field.
+  """
+  @doc since: "0.6.0"
+  @spec validate_datetime_order(Ecto.Changeset.t(), atom, atom, list(atom), fun) ::
+          Ecto.Changeset.t()
+  def validate_datetime_order(
+        changeset,
+        from_field,
+        until_field,
+        valid_orders \\ [:lt, :eq],
+        formatter \\ &to_local_datetime/2
```

Comment:
🤔  hmm this `to_local_datetime` thing is a bit odd, IMO. If you absolutely want to include it, maybe move it to `BitcrowdEcto.DateTime`... But perhaps just default to `DateTime.to_string`? It seems oddly specific with the timezone shift, etc.

Reply:
Indeed, and that's solved by turning the formatter into a simpler form in `validate_order` 👍